proposals/p1885.md

```diff
@@ -0,0 +1,864 @@
+template <typename Iterate>
+class IterateIterator {
+ public:
+  // Returns an "Invalid" iterator representing `end()`
+  static auto Invalid(const Iterate& container) -> IterateIterator<Iterate> {
+    return IterateIterator<Iterate>(container, std::nullopt);
+  }
+
+  IterateIterator(const Iterate& container,
+                  std::optional<typename Iterate::CursorType> cursor)
+      : container_(&container), cursor_(cursor) {
+    if (cursor_) {
+      next();
+    }
+  }
+  IterateIterator(const IterateIterator& other)
+      : container_(other.container_), cursor_(other.cursor_),
+      value_(other) {}
+
+  auto operator*() const -> const typename Iterate::ElementType& {
+    assert(cursor_);
+    return value_;
+  }
+  auto operator++() -> IterateIterator<Iterate>& {
+    next();
+    return *this;
+  }
+  auto operator==(const IterateIterator<Iterate>& rhs) -> bool {
+    return container_ == rhs.container_ && cursor_ == rhs.cursor_;
+  }
+  auto operator!=(const IterateIterator<Iterate>& rhs) -> bool {
+    return !(*this == rhs);
+  }
+
+ private:
+  void next() {
+    assert(cursor_);
+    if (const auto v = container_->Next(*cursor_); v) {
+      value_ = *std::move(v);
+    } else {
+      cursor_ = std::nullopt;
+    }
+  }
+
+  const Iterate* container_;
+  std::optional<typename Iterate::CursorType> cursor_;
+  typename Iterate::ElementType value_;
+};
+
+// A type implementing `Iterate` with CursorType `C`, and
+// ElementType `T` could have `begin()` and `end()` methods
+// defined as such:
+class IterateType ... {
+  auto begin() -> IterateIterator<Iterable<C, T>> {
+    return IterateIterator<Iterable<C, T>>(*this, NewCursor());
+  }
+  auto end() -> IterateIterator<Iterable<C, T>> {
+    return IterateIterator<Iterable<C, T>>::Invalid(*this);
+  }
+};
+}  // namespace Carbon
+
+// Usage:
+Carbon::MyContainer container;
+for (auto s : container) { ...}
+```
+
+### Inversion of control
+
+Using an `Optional` has downsides, discussed in the
+["alternatives considered"](#alternatives-considered) section. A possible
+approach would be to invert the control of the `for` loop, by having the
+container call in the `for` block for each value, passing the value as argument.
+This would work similarly to [Python generator functions](#python).
+
+This has the following advantages:
+
+-   Removes the need for an `Optional`, and the associated overheads, copies,
+    and unwrapping.
+-   No boundary checks needed at the `for` level
+-   Compatible with R-value containers
+
+Limitations:
+
+-   The context needed by the `for` body will needs to be available or provided
+-   No optimizer known to the team can reliably produce efficient code for this
+    construct.
+
+## Rationale
+
+This proposal offers a first step to support a needed way to iterate using a
+`for` loop. More specifically:
+
+-   Requires a single interface to be implemented with a combined `Next` method,
```

Comment:
It might also be worth mentioning that a combined `Next` method makes it easier to support ranges that consume their input. For example, in C++20 if you filter an `istream_view` through `take(5)` (which filters a range to its first 5 elements), the first *6* elements of the stream are consumed, because the C++ iterator model forces `istream_iterator` to consume an element on initialization (so that the first `operator*` has a result to return) and on every `operator++` call. As another example of the same kind of problem, `iota(0) | filter([](auto x) { return x < 5; })) | take(5)` is an infinite loop. I think a combined `Next` method will avoid these problems.

I got this from a talk by Barry Rezvin at C++Now last week; I'm not sure when the video will be online, but you can see the slides [here](https://schedule.cppnow.org/wp-content/uploads/2023/03/take5-slides.pdf)

Reply:
Added to the "Alternatives" section, as this seemed to best belong there.